File: tests/test_client.py
"""Unit tests for the Sorted Travel REST client."""

import unittest

from sorted_travel.client import APIError, Client


class ClientTests(unittest.TestCase):
    """Cover URL building, auth headers, and error mapping."""

    def test_status_uses_public_path(self) -> None:
        """status() GETs /api/v1/status with no Authorization header."""
        captured = {}

        def transport(request, timeout):
            captured["request"] = request
            captured["timeout"] = timeout
            return 200, "application/json", '{"status":"OK"}'

        client = Client(transport=transport, env={})
        payload = client.status()
        self.assertEqual(payload, {"status": "OK"})
        self.assertEqual(
            captured["request"]["url"],
            "https://sorted.travel/api/v1/status",
        )
        self.assertEqual(captured["request"]["method"], "GET")
        self.assertNotIn("authorization", captured["request"]["headers"])

    def test_list_destinations_encodes_cursor(self) -> None:
        """list_destinations() passes cursor and limit as query params."""
        captured = {}

        def transport(request, timeout):
            del timeout
            captured["url"] = request["url"]
            return 200, "application/json", '{"data":[],"has_more":false}'

        client = Client(transport=transport, env={})
        client.list_destinations(cursor="abc+1", limit=5)
        self.assertIn("cursor=abc%2B1", captured["url"])
        self.assertIn("limit=5", captured["url"])

    def test_api_key_sets_bearer_header(self) -> None:
        """A configured key is sent as Authorization: Bearer."""
        captured = {}

        def transport(request, timeout):
            del timeout
            captured["headers"] = request["headers"]
            return 200, "application/json", "{}"

        client = Client(api_key="st_sandbox_test", transport=transport, env={})
        client.status()
        self.assertEqual(
            captured["headers"]["authorization"],
            "Bearer st_sandbox_test",
        )

    def test_non_2xx_raises_api_error(self) -> None:
        """HTTP errors surface as APIError with status and body."""

        def transport(request, timeout):
            del request, timeout
            return 429, "application/json", '{"error":"rate_limited"}'

        client = Client(transport=transport, env={})
        with self.assertRaises(APIError) as raised:
            client.status()
        self.assertEqual(raised.exception.status, 429)
        self.assertEqual(raised.exception.body, {"error": "rate_limited"})


if __name__ == "__main__":
    unittest.main()

File: ruby/test/test_client.rb
# frozen_string_literal: true

require "minitest/autorun"

require_relative "../lib/sorted_travel"

class ClientTest < Minitest::Test
  def test_status_uses_public_path
    captured = nil
    client = SortedTravel::Client.new(
      env: {},
      transport: lambda do |url:, method:, headers:, body:, timeout:|
        captured = { url: url, method: method, headers: headers, body: body }
        [200, "application/json", '{"status":"OK"}']
      end,
    )
    payload = client.status
    assert_equal({ "status" => "OK" }, payload)
    assert_equal("https://sorted.travel/api/v1/status", captured[:url])
    assert_equal("GET", captured[:method])
    assert_nil(captured[:headers]["Authorization"])
  end

  def test_list_destinations_encodes_cursor
    captured_url = nil
    client = SortedTravel::Client.new(
      env: {},
      transport: lambda do |url:, method:, headers:, body:, timeout:|
        captured_url = url
        [200, "application/json", '{"data":[],"has_more":false}']
      end,
    )
    client.list_destinations(cursor: "abc+1", limit: 5)
    assert_match(/cursor=abc%2B1/, captured_url)
    assert_match(/limit=5/, captured_url)
  end

  def test_api_key_sets_bearer_header
    captured_headers = nil
    client = SortedTravel::Client.new(
      api_key: "st_sandbox_test",
      env: {},
      transport: lambda do |url:, method:, headers:, body:, timeout:|
        captured_headers = headers
        [200, "application/json", "{}"]
      end,
    )
    client.status
    assert_equal("Bearer st_sandbox_test", captured_headers["Authorization"])
  end

  def test_non2xx_raises_api_error
    client = SortedTravel::Client.new(
      env: {},
      transport: lambda do |url:, method:, headers:, body:, timeout:|
        [429, "application/json", '{"error":"rate_limited"}']
      end,
    )
    error = assert_raises(SortedTravel::APIError) { client.status }
    assert_equal(429, error.status)
    assert_equal({ "error" => "rate_limited" }, error.body)
  end
end
